.github/workflows/ci.yaml

```diff
@@ -1,44 +1,67 @@
-name: ci
+name: CI
 
-on:
-  push:
-    branches: [ "main" ]
-  pull_request:
-    branches: [ "main" ]
+on: ["push", "pull_request"]
 
 permissions:
   contents: read
 
 jobs:
-  build:
+  test:
+    name: "Python ${{ matrix.python-version }} on ${{ matrix.os }}"
+    runs-on: "${{ matrix.os }}"
 
-    runs-on: ubuntu-latest
     strategy:
+      fail-fast: false
       matrix:
-        python-version: ["3.9", "3.10", "3.11", "3.12"]
+        os:
+          - ubuntu-latest
+          - macos-latest
+          - windows-latest
+        python-version:
+          - "3.9"
+          - "3.10"
+          - "3.11"
+          - "3.12"
+          - "3.13"
 
     steps:
-      - uses: actions/checkout@v3
-      - name: Set up Python
-        uses: actions/setup-python@v4
+      - name: "Check out the repository"
+        uses: actions/checkout@v4
+
+      - name: "Set up Python ${{ matrix.python-version }}"
+        uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.python-version }}
           cache: "pip"
           cache-dependency-path: pyproject.toml
-      - uses: actions/cache@v3
-        id: cache
-        with:
-          path: ${{ env.pythonLocation }}
-          key: ${{ runner.os }}-python-${{ env.pythonLocation }}-${{ hashFiles('pyproject.toml') }}-test-v03
-      - name: Install Dependencies
-        if: steps.cache.outputs.cache-hit != 'true'
+
+      - name: "Install dependencies"
         run: |
           python -m pip install --upgrade pip
           pip install -e .
           pip install -r requirements-dev.txt
-      - name: run pytest py${{ matrix.python-version }}
-        run: |
-          make test
-      - name: run tox codestyle
-        run: |
-          make check-codestyle
+
+      - name: "Check codestyle"
+        run: make check-codestyle
+
+      - name: "Run type checker"
+        run: make check-types
+
+      - name: "Run tests & coverage"
+        run: make test
+
+      - name: "Post code coverage data to Coveralls"
+        uses: coverallsapp/github-action@v2
+        with:
+          flag-name: coverage-${{ join(matrix.*, '-') }}
+          parallel: true
+
+  finish:
+    needs: test
+    if: ${{ always() }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: "Coveralls Finished"
```

Comment:
this notifies Coveralls that all parallel jobs have finished & their respective output coverages uploaded